source/common/upstream/upstream_impl.cc

```diff
@@ -1098,8 +1159,27 @@ ClusterInfoImpl::ClusterInfoImpl(
   }
 
   // If load_balancing_policy is set we will use it directly, ignoring lb_policy.
-  if (config.has_load_balancing_policy()) {
+  if (config.has_load_balancing_policy() ||
+      config.lb_policy() == envoy::config::cluster::v3::Cluster::LOAD_BALANCING_POLICY_CONFIG) {
     configureLbPolicies(config, server_context);
+  } else if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.convert_legacy_lb_config")) {
+    auto lb_pair = getTypedLbConfigFromLegacyProto(config);
+    load_balancer_config_ = std::move(lb_pair.first);
+    load_balancer_factory_ = lb_pair.second;
+    lb_type_ = LoadBalancerType::LoadBalancingPolicyConfig;
+
+    RELEASE_ASSERT(
+        load_balancer_factory_,
+        fmt::format(
+            "No load balancer factory found from legacy LB configuration (type: {}, subset: {}).",
+            static_cast<int>(config.lb_policy()), config.has_lb_subset_config()));
+
+    // Clear unnecessary legacy config because all legacy config is wrapped in load_balancer_config_
+    // except the original_dst_lb_config.
+    lb_subset_ = nullptr;
+    if (config.lb_policy() != envoy::config::cluster::v3::Cluster::CLUSTER_PROVIDED) {
+      lb_policy_config_ = nullptr;
```

Comment:
nit, it seems ok to not cleanup this, then you needn't all the changes in the upstream_impl.h

Reply:
This is for reducing unnecessary memory overhead. We have lots of related PR to reduce memory overhead of clusters. I don't want to break their achievements.
Although I am open to this problem because all these code will be removed after six months.